distributed/scheduler.py

```diff
@@ -5648,143 +5650,172 @@ async def gather(self, keys, serializers=None):
+            raise ValueError(
+                (
+                    "Expected all workers to have a nanny, encountered "
+                    f"{len(non_nannies)} worker(s) without a nanny."
+                ),
+                non_nannies,
+            )
+
+    @log_errors
+    async def restart_workers(self, workers: Iterable[str], timeout: float = 30):
+        """
+        Restart a specified set of workers
+
+        This methods expects all workers to have nannies to be able to restart them.
+        If workers without nannies exist, ``Client.restart_workers`` will return a
+        serialized ``ValueError`` that lists the workers without nannies. Consider
+        removing those workers and calling ``Client.restart_workers`` again afterward.
+
+        Parameters
+        ----------
+        workers:
+            Workers to restart
+        timeout:
+            Return a `RuntimeError` if ``restart_workers`` takes more than ``timeout``
+            seconds
 
-        logger.debug("Send kill signal to nannies: %s", nanny_workers)
+        See Also
+        --------
+        Scheduler.restart
+        Client.restart_workers
+        """
+        try:
+            self._expect_nannies(workers)
+            logger.debug("Asking nannies to restart workers: %s.", workers)
+            await self._restart_workers(workers, timeout)
+            return {"status": "OK"}
+        except RuntimeError as e:
+            logger.exception(e)
+            return error_message(e)
+
+    async def _restart_workers(self, workers: Iterable[str], timeout: float) -> None:
+        nanny_addresses = [self.workers[worker].nanny for worker in workers]
         async with contextlib.AsyncExitStack() as stack:
             nannies = await asyncio.gather(
                 *(
                     stack.enter_async_context(
-                        rpc(nanny_address, connection_args=self.connection_args)
+                        rpc(address, connection_args=self.connection_args)
                     )
-                    for nanny_address in nanny_workers.values()
+                    for address in nanny_addresses
                 )
             )
 
-            start = monotonic()
-            resps = await asyncio.gather(
+            responses = await asyncio.gather(
                 *(
                     asyncio.wait_for(
-                        # FIXME does not raise if the process fails to shut down,
-                        # see https://github.com/dask/distributed/pull/6427/files#r894917424
-                        # NOTE: Nanny will automatically restart worker process when it's killed
-                        nanny.kill(
-                            reason="scheduler-restart",
-                            timeout=timeout,
-                        ),
-                        timeout,
+                        nanny.restart(timeout=timeout, reason="scheduler-restart"),
+                        timeout=timeout,
                     )
```

Comment:
FYI these tasks will raise a CommClosedError if a worker is disconnecting from the scheduler while the restart is running.

We're calling `ConnectionPool.remove` for the worker that is being removed which closes all connection to that worker, including the one we're using here. This might be the exception we're seeing in https://github.com/coiled/coiled-runtime/issues/468#issuecomment-1309528389
i.e. this would likely happen with or without `wait_for_workers`. I'm wondering though why this is isn't happening even more often

Reply:
Here we're connecting to the nanny addresses, not the workers. The nanny connection shouldn't be closed by restart, so I wouldn't expect this to raise `CommClosedError`